### ISSUE TITLE
use different class weights for training

### ISSUE THREAD

**Issue by shivahanifi** [shivahanifi/SCDD-image-segmentation-keras#2]
Specific class weights are focused on five classes in the dataset: background, ribbons, cracks, gridline defects, and inactive areas.

A two-level screening design was selected for the experiment because it is ideal for exploring many factors with a minimum of experimental runs. Each factor is set to a high and a low value to examine the main effects and interactions. The low values were set to 50% of the custom class weights, and the high values were set to 50% of the inverse class weights.

![Screenshot from 2024-10-22 11-13-36](https://github.com/user-attachments/assets/66fec5bc-b0e2-421a-9b48-3b0e866383ff)

**Comment by shivahanifi** [comment]
For handling class imbalance, you can pass the class weights directly to the loss function. The keras_segmentation library doesn’t natively support custom class weights in the loss function, so you need to implement a custom loss function that includes these weights.
1. Defining class weights
    manually or using the sklearn function compute_class_weight
2. customize the loss function
e.g.

```
import keras.backend as K

def weighted_categorical_crossentropy(weights):
    def loss(y_true, y_pred):
        y_true = K.one_hot(K.cast(K.flatten(y_true), 'int32'), num_classes=len(weights))
        y_pred = K.flatten(y_pred)
        # Calculate weighted loss
        loss = K.categorical_crossentropy(y_true, y_pred)
        loss = K.sum(loss * K.constant(weights))
        return loss
    return loss

# Define the loss function with the computed class weights
custom_loss = weighted_categorical_crossentropy(list(class_weight_dict.values()))
```
3. Compile the Model with Custom Loss
```
# Re-compile the model with the custom loss function
model.compile(optimizer='adam', loss=custom_loss, metrics=['accuracy'])

# Now you can proceed with training
model.train(
    train_images=train_image_path,
    train_annotations=train_annotations_path,
    checkpoints_path=checkpoint_path,
    epochs=wandb.config.epochs,
    batch_size=2,
    steps_per_epoch=len(os.listdir(train_image_path)) // 2,
    callbacks=[WandbCallback(), checkpoint_callback]
)
```
4. Log to wandb
```
# Log class weights to WandB
wandb.config.update({
    "class_weights": class_weight_dict
})

```

**Comment by shivahanifi** [comment]
## Using `loss_weights` option when compiling

sing `loss_weights` in `model.compile()` does not accomplish the same thing as using class weights in a segmentation task. 
- The `loss_weights` parameter in `model.compile()` is used to balance multiple output losses in models with more than one output. It assigns weights to the different outputs (e.g., if your model has two outputs, like two separate segmentation maps or classifications). It doesn’t affect individual class weights within a single loss function.
- In contrast, class weights are used within a single output loss to adjust the **importance of each class** in the loss calculation. This is particularly important in segmentation tasks where certain classes are underrepresented and need to be "weighted up" to avoid imbalance in learning.

check [link](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) for detailed info.

**Comment by shivahanifi** [comment]
## Implementation with no error (did not compile successfully)
```

# class weight assignment - set C
weights = [
    0.2,  # Label 0: "bckgnd" 
    1.0,  # Label 1: "sp multi" 
    1.0,  # Label 2: "sp mono" 
    1.0,  # Label 3: "sp dogbone" 
    3.0,  # Label 4: "ribbons"
    1.0,  # Label 5: "border" 
    1.0,  # Label 6: "text" 
    1.0,  # Label 7: "padding" 
    1.0,  # Label 8: "clamp" 
    1.0,  # Label 9: "busbars" 
    1.0,  # Label 10: "crack rbn edge" 
    10.0, # Label 11: "inactive" 
    1.0,  # Label 12: "rings" 
    1.0,  # Label 13: "material" 
    20.0, # Label 14: "crack" 
    10.0, # Label 15: "gridline" 
    1.0,  # Label 16: "splice" 
    1.0,  # Label 17: "dead cell"
    1.0,  # Label 18: "corrosion" 
    1.0,  # Label 19: "belt mark" 
    1.0,  # Label 20: "edge dark" 
    1.0,  # Label 21: "frame edge" 
    1.0,  # Label 22: "jbox" 
    1.0,   # Label 23: "meas artifact"
]

# Log class weights to WandB
wandb.config.update({
    "class_weights": weights
})

# Define weighted categorical cross-entropy loss
def weighted_categorical_crossentropy(weights):
    def loss(y_true, y_pred):
        y_true = K.one_hot(K.cast(K.flatten(y_true), 'int32'), num_classes=len(weights))
        y_pred = K.flatten(y_pred)
        # Calculate weighted loss
        loss = K.categorical_crossentropy(y_true, y_pred)
        loss = K.sum(loss * K.constant(weights))
        return loss
    return loss

# Define the loss function with the computed class weights
custom_loss = weighted_categorical_crossentropy(weights)

# Define the model 
model = vgg_unet(n_classes=wandb.config.n_classes ,  input_height=wandb.config.input_height, input_width=wandb.config.input_width)

# Re-compile the model with the custom loss function
model.compile(optimizer='adam', loss=custom_loss, metrics=['accuracy'])
```

**Comment by shivahanifi** [comment]
## TO DO 1: compare the performance

The performance was the same for the model without specific weights ([train_SCDD_20211104_augmented_e1_speFull](https://wandb.ai/ubix/scdd_segmentation_keras/runs/uu6c7f0h?nw=nwusershivahanifi)) and the one with custom weights  ([train_SCDD_20211104_w3_e1_speFull_v2](https://wandb.ai/ubix/scdd_segmentation_keras/runs/a1woj1sx?nw=nwusershivahanifi)) applied.

## ISSUE
The model is not compiling correctly and it is actually using the previous categorical_crossentropy loss instead of the weighted_categorical_crossentropy.

**Comment by shivahanifi** [comment]
## TO DO 2: check the loggings
## TO DO 3: check if the weights are applied also in metrics

The code is not applying the weighted loss.

**Comment by shivahanifi** [comment]
## [Weighted Categorical Cross-Entropy loss object definition](https://medium.com/the-owl/weighted-categorical-cross-entropy-loss-in-keras-edaee1df44ee)

Followin this article to implement the weighted loss in keras.

## Solved: 

Using the function and class defined in the article and implementing it in the `train.py` script solved the problem.
```
def weighted_categorical_crossentropy(target, output, weights, axis=-1):
    target = tf.convert_to_tensor(target)
    output = tf.convert_to_tensor(output)
    target.shape.assert_is_compatible_with(output.shape)
    weights = tf.reshape(tf.convert_to_tensor(weights, dtype=target.dtype), (1,-1))

    # Adjust the predictions so that the probability of
    # each class for every sample adds up to 1
    # This is needed to ensure that the cross entropy is
    # computed correctly.
    output = output / tf.reduce_sum(output, axis, True)

    # Compute cross entropy from probabilities.
    epsilon_ = tf.constant(tf.keras.backend.epsilon(), output.dtype.base_dtype)
    output = tf.clip_by_value(output, epsilon_, 1.0 - epsilon_)
    return -tf.reduce_sum(weights * target * tf.math.log(output), axis=axis)


#@tf.keras.saving.register_keras_serializable(name="WeightedCategoricalCrossentropy")
class WeightedCategoricalCrossentropy:
    def __init__(
        self,
        weights,
        label_smoothing=0.0,
        axis=-1,
        name="weighted_categorical_crossentropy",
        fn = None,
    ):
        """Initializes `WeightedCategoricalCrossentropy` instance.

        Args:
          weights: A tensor or array of weights to apply to each class. 
          This can be helpful when you have an imbalanced dataset and want 
          to give more importance to certain classes.
          label_smoothing: Float in [0, 1]. When 0, no smoothing occurs. When >
            0, we compute the loss between the predicted labels and a smoothed
            version of the true labels, where the smoothing squeezes the labels
            towards 0.5.  Larger values of `label_smoothing` correspond to
            heavier smoothing.
          axis: The axis along which to compute crossentropy (the features
            axis).  Defaults to -1.
          name: Name for the op. Defaults to 'weighted_categorical_crossentropy'.
          fn: If provided, this will override the default function, 
          allowing custom behavior in case a different function is passed.
        """
        super().__init__()
        self.weights = weights # tf.reshape(tf.convert_to_tensor(weights),(1,-1))
        self.label_smoothing = label_smoothing
        self.name = name
        self.fn = weighted_categorical_crossentropy if fn is None else fn

    def __call__(self, y_true, y_pred, axis=-1):
        if isinstance(axis, bool):
            raise ValueError(
                "`axis` must be of type `int`. "
                f"Received: axis={axis} of type {type(axis)}"
            )
        y_pred = tf.convert_to_tensor(y_pred)
        y_true = tf.cast(y_true, y_pred.dtype)
        self.label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)

        if y_pred.shape[-1] == 1:
            warnings.warn(
                "In loss categorical_crossentropy, expected "
                "y_pred.shape to be (batch_size, num_classes) "
                f"with num_classes > 1. Received: y_pred.shape={y_pred.shape}. "
                "Consider using 'binary_crossentropy' if you only have 2 classes.",
                SyntaxWarning,
                stacklevel=2,
            )

        def _smooth_labels():
            num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
            return y_true * (1.0 - self.label_smoothing) + (self.label_smoothing / num_classes)

        y_true = tf.__internal__.smart_cond.smart_cond(self.label_smoothing, _smooth_labels, lambda: y_true)

        return tf.reduce_mean(self.fn(y_true, y_pred, self.weights, axis=axis))
    
    def get_config(self):
        config = {"name":self.name, "weights": self.weights, "fn": weighted_categorical_crossentropy}

        # base_config = super().get_config()
        return dict(list(config.items()))

    @classmethod
    def from_config(cls, config):
        """Instantiates a `Loss` from its config (output of `get_config()`).

        Args:
            config: Output of `get_config()`.
        """
        if saving_lib.saving_v3_enabled():
            fn_name = config.pop("fn", None)
            if fn_name: # and cls is LossFunctionWrapper:
                config["fn"] = get(fn_name)
        return cls(**config)
```

**Comment by shivahanifi** [comment]
closing as complete. Follow on the issue #23
